### DISCUSSION TITLE
Why .gitignore meta/src/grammar.rs?

### DISCUSSION THREAD

**Discussion by NoahTheDuke** [pest-parser/pest#558] [Q&A] [answered]
As I understand it so far (still learning the codebase), grammar.rs is generated by the `bootstrap` bin. Seeing as `bootstrap` relies on a trusted version of pest, why not include the grammar.rs file directly? Developers will still have to run `bootstrap` every time they change the base `grammar.pest` file, but including the generated file will allow for manually verifying (with git diff, etc) that a given change works how one expects.

**Comment by CAD97** [comment, marked as answer]
The general reason is that generally, [you shouldn't put generated files in source control](https://softwareengineering.stackexchange.com/a/192115/222879).

But more generally, the bootstrap system is quite messy. Previously, pest@2 used pest@1 for its grammar, IIRC. I'm the one who set up the current bootstrap to avoid the cyclic requirement. The ultimate goal is that crates-io users don't have to build more than one pest, but we can still dogfood by using crates-io pest to develop the next release of pest.

I've played on-and-(mostly)-off with a way of "properly" solving the bootstrap problem (specifically for the case of using watt to drive proc macro crates). The ideal is that `cargo build` just works, but more importantly, that the two library-user usecases just work:

- crates-io dependency just gets the already built bootstrapped generated code (for watt, the wasm; for codegen, the codegenned Rust files), and
- git dependencies continue to just work, generating the generated files.

For a binary blob like wasm, that changes any time anything in the macro impl is changed, not committing it to the repo is more important. r-a is a notable example of tracking generated source files in source control, and their use case likely changes around the same amount as the pest grammar would change.

I'm not strongly against committing the generated source file again, so long as [it's marked as generated](https://github.com/github/linguist/blob/master/docs/overrides.md#generated-code) and [a test to make sure it's up-to-date](https://github.com/rust-analyzer/rust-analyzer/blob/86c534f244bd72dc9e9ec1be0bfdf0131602ca30/crates/sourcegen/src/lib.rs#L166-L189).

**Reply by NoahTheDuke** [reply]
That all makes sense. For me, the grammar.rs file feels more akin to a Cargo.lock file we want to keep explicitly in sync with the "human readable"/canonical file (grammar.pest).

I'm still figuring out how this stuff works so I'll defer to your expertise here lol.
